## Results {.page_break_before}

Hackmd for tables here: https://hackmd.io/GvngZ4gHQE-9ERB4Gd71HQ

### Many individual k-mers are genome specific

Consider a collection of the size of GTDB - 320,000 genomes with
an average size of 5e6 bp. In the simplest case where these genomes
were completely random and had neither bias nor redundancy, 
most 31-mers present in one genome should be distinct to that genome.
Thus the presence of any one k-mer in a sequencing data set
should be capable of perfectly distinguishing that genome's presence
even in low-coverage data sets
(Appendix A).

In practice, genomes are neither unbiased nor contain distinct
content.  We used FracMinHash (S=1000, k=31) to generate a
representative collection of hashes and found that 15378449 of
22792206 hashvals (67.5%) in GTDB rs207 are perfectly informative at
genome level.

K-mers can be further ranked based on *how much* discriminatory power they
offer - for example, a k-mer that is only present in two genomes in the
collection is much more informative than a k-mer present in 20. One way to
formalize this is with Shannon entropy (formula, base 2).

| # genomes containing | H (Shannon entropy) | # hashes | % total |
| -------- | -------- | -------- | -- |
| 1 | 0 | 15,378,449 | 67.5% |
| 2 | 1 | 3,407,595 | 15.0% |
| 3 | 1.58 | 1,337,246 | 5.9%|
| 4 | 2 | 694,687 | 3.0% |
| 5 or more | >= 2.32 | 1,974,229 | 8.7% |

Table: Distribution of 31-mers among distinct genomes in GTDB rs207,
estimated with FracMinHash (scaled=1000). {#tbl:genome-entropy}

What is value of H here, exactly? Other than H=0? Link to taxonomy, and
combinatorics; also see unicity distance at bottom.

These 15.4m perfectly discriminating k-mers are not equally
distributed across genomes, however. Of the 318k genomes in GTDB
rs207, only 149k (47.1%) of the genomes contain a perfect k-mer,
suggesting that slightly more than half of the genomes cannot be
classified accurately when using any one k-mer.

(CTB: be careful about estimates based on FracMinHash, vs actual k-mers.)

### Shannon entropy of k-mers can be used to measure taxonomic informativeness

We measured the species distribution in GTDB rs207 for 21.2 million
hashes, representing 21.2 billion 31-mers, and calculated the Shannon
entropy for each hash (equationXX) at the species, genus, and family
levels. Per Table @tbl:gtdb-entropy, 99.1% of hashes uniquely identify
a specific family within the GTDB taxonomy.

| Taxonomic level (GTDB) | # perfectly informative hashes | cumulative total % |
| -------- | -------- | -------- |
| species | 21,150,287 | 92.8% |
| genus | 1,262,281 | 98.3% |
| family | 170,249 | 99.1% |
| order and above | 209,389 | 0.9% |

Table: Entropy measurements for GTDB taxonomy using 318k genomes from
rs207 genomes. {#tbl:gtdb-entropy}

Make point that nucleotide k-mers are not necessarily specific beyond
family (ref protein paper).

Add fourth row to tables - rest?

Discuss entropy and genome-specific. Transition to combinatorics?

### Shannon entropy can summarize the taxonomic cohesion of taxonomies based on genomic relationships.

We can also calculate the Shannon entropy with respect to different
taxonomies. Table @tbl:ncbi-entropy uses the NCBI taxonomy with the
same genomes used above.  Here we see that approximately 4.5% of hashes
cannot be used to distinguish between different families
- a full 5 times as many as with the GTDB taxonomy.  These
1.0 million hashes represent approximately 10 billion k-mers, or
approximately 2,000 bacterial genomes worth of sequence.

| Taxonomic level (NCBI) | # perfectly informative hashes | cumulative total % |
| -------- | -------- | -------- |
| species | 20,744,791 | 91.0% |
| genus | 779,234 | 94.4% |
| family | 245,718 | 95.5% |
| order and above | 1,022,463 | 4.5% |

Table: Entropy measurements for same 318k GTDB rs207 genomes as in
Table @tbl:gtdb-entropy, but using NCBI taxonomic
labels. {#tbl:ncbi-entropy}

### Many k-mers with non-zero high-tax entropy come from a few specific genomes

Explore taxonomic incoherence and database contamination.

Find challenging genomes. Explore low H values.

### Unicity distance can be used to estimate genomic redundancy

We next ask, how many genomes can be distinguished from each other
using a combinatorial collection of k-mers?  To do this, we estimate
the _unicity distance_ of each genome in the database, where the
unicity distance is defined as the smallest set of hashes capable of
uniquely identifying an individual genome. (k=31, scaled=1000)

Table @tbl:unicity shows that approximately 29.2% of the genomes in
GTDB rs207 cannot be distinguished uniquely by _any_ combination of
31-mers at a scaled of 1000, while some substantial amount (estimated
as 15.3%) can be precisely identified using a single hash.

(Do higher resolution k-mer analysis of some of these; talk about
error rates, etc.)

(Compare also with k-mer informativeness; can we tie entropy
computation at top back to number of genomes with unicity of 1,
and cross validate?)
 
| Unicity distance | Number of genomes | Percent of genomes |
| -------- | -------- | -------- |
| 1     | 48,630     | 15.3%     |
| infinite | 92,564 | 29.2% | 

Table: Estimated unicity distances with hashes for 318k GTDB rs207
genomes using FracMinHash as implemented in sourmash (k=31,
scaled=1000). {#tbl:unicity}
