## Results {.page_break_before}

### Many k-mers are genome specific

### Shannon entropy of k-mers can be used to measure taxonomic informativeness

We measured the species distribution in GTDB rs207 for 21.2 million hashes,
representing 21.2 billion 31-mers, and calculated the Shannon entropy of
species for each hash (equationXX). Per Table XX, 92.8% of hashes uniquely
identify a specific family.

| Taxonomic level | # perfectly informative hashes | cumulative % total | 
| -------- | -------- | -------- |
| species     | 21,150,287     | 92.8%     | 92.8%
| genus     | 1,262,281     | XXX%     |
| family     | 170,249     | YYY%     | 

(Make point that nucleotide k-mers are not necessarily specific beyond family ref protein paper.)

### Many k-mers with non-zero entropy come from a few specific genomes

Explore taxonomic incoherence and database contamination.

### Unicity distance can be used to estimate genomic redundancy

We next ask, how many genomes can be distinguished from each other
using a combinatorial collection of k-mers?  To do this, we estimate
the _unicity distance_ of each genome in the database, where the
unicity distance is defined as the smallest set of hashes capable of
uniquely identifying a genome. (k=31, scaled=1000)

Table YY shows that approximately 29.2% of the genomes in GTDB rs207
cannot be distinguished uniquely by any combination of k-mers with
these parameters.
 
| Unicity distance | Number of genomes | Percent of genomes |
| -------- | -------- | -------- |
| 1     | 48,630     | 15.3%     |
| infinite | 92,564 | 29.2% | 
