## Results {.page_break_before}

Hackmd for tables here: https://hackmd.io/GvngZ4gHQE-9ERB4Gd71HQ

### Many k-mers are genome specific

### Shannon entropy of k-mers can be used to measure taxonomic informativeness

We measured the species distribution in GTDB rs207 for 21.2 million
hashes, representing 21.2 billion 31-mers, and calculated the Shannon
entropy for each hash (equationXX) at the species, genus, and family
levels. Per Table @tbl:gtdb-entropy, 99.1% of hashes uniquely identify
a specific family within the GTDB taxonomy.

| Taxonomic level (GTDB) | # perfectly informative hashes | cumulative % total | 
| -------- | -------- | -------- |
| species     | 21,150,287     | 92.8%     | 92.8%
| genus     | 1,262,281     | 98.3%     |
| family     | 170,249     | 99.1%     | 

Table: Entropy measurements for GTDB taxonomy using 318k genomes from
rs207 genomes. {#tbl:gtdb-entropy}

Make point that nucleotide k-mers are not necessarily specific beyond
family (ref protein paper).

### Shannon entropy can summarize the taxonomic cohesion of taxonomies based on genomic relationships.

We can also calculate the Shannon entropy with respect to different
taxonomies. Table @tbl:ncbi-entropy uses the NCBI taxonomy with the
same genomes used above.  Here we see that approximately 4.5% of hashes
cannot be used to distinguish between different families
- a full 5 times as many as with the GTDB taxonomy.  These
1.0 million hashes represent approximately 10 billion k-mers, or
approximately 2,000 bacterial genomes worth of sequence.

| Taxonomic level (NCBI) | # perfectly informative hashes | cumulative % total | 
| -------- | -------- | -------- |
| species     | 20,744,791    | 91.0% |
| genus     | 779,234    | 94.4%     |
| family     | 245,718    | 95.5%     | 

Table: Entropy measurements for same 318k GTDB rs207 genomes as in
Table @tbl:gtdb-entropy, but using NCBI taxonomic
labels. {#tbl:ncbi-entropy}

### Many k-mers with non-zero entropy come from a few specific genomes

Explore taxonomic incoherence and database contamination.

### Unicity distance can be used to estimate genomic redundancy

We next ask, how many genomes can be distinguished from each other
using a combinatorial collection of k-mers?  To do this, we estimate
the _unicity distance_ of each genome in the database, where the
unicity distance is defined as the smallest set of hashes capable of
uniquely identifying a genome. (k=31, scaled=1000)

Table YY shows that approximately 29.2% of the genomes in GTDB rs207
cannot be distinguished uniquely by any combination of 31-mers at a
scaled of 1000.

(Do higher resolution k-mer analysis of some of these; talk about
error rates, etc.)
 
| Unicity distance | Number of genomes | Percent of genomes |
| -------- | -------- | -------- |
| 1     | 48,630     | 15.3%     |
| infinite | 92,564 | 29.2% | 

