## Discussion {.page_break_before}

### Species-level classification should be straightforward with k-mers

Taxonomic classification to the species level is straightforward,
largely because GTDB taxonomy is closely tied to genomic content and
most of the genomic redundancy lies within species and genus
level. Thus GTDB taxonomy largely encapsulates this redundancy. The
entropy measurements demonstrate that it is possible to choose an
informative subset of k-mers that would robustly classify at the
species level, and that doing so would not compromise sensitivity.
LCA-style approaches such as those used by Kraken should work even if
we use genus and family level k-mers, while eliminating those above.

Shared genomic content at higher levels confounds taxonomic
classification methods. While surely some shared genomic content is
real, our analysis suggests that significant portions of it are
contamination.

### Classification below the strain level

Detecting genomes from sequences is easy with k-mers, but
significant redundancy prevents straightforward classification to the
genome/strain level. Here leveraging combinatorial application of
k-mers provides significant leverage; this is how sourmash achieves
high precision. Nonetheless sourmash cannot distinguish a full 30% of
the genomes in the database from each other.

Some implications are that it should be possible to use information
from both short and long reads to classify robustly to the species
level, but it is unlikely to work below that (at the strain level).
This is because many reads will map to shared content within a species,
and some of that shared content may not distinguish a particular
genome from others in the pangenome at the resolution of the available
reads.

A simple thought experiment also suggests that reduced-representation
/slimmed-down databases will not support strain-level classification.
Suppose that a technique exists that can classify reads to a strain
level.  First, choose a read that belongs to two or more different
strains; there is no way to identify which strain this belongs
to. Second, choose a read that belongs to a strain that is not
represented in the database; while it clearly belongs to a known
species, there is no way to identify which. Classifiers should be
using all available information and it is clearly possible to do so,
viz sourmash.

(Probably need to spend some time here talking about core vs accessory
genomes.)

Approaches such as sourmash can try to operate "above" individual
k-mers, but will also be stymied by infinite unicity. Here combinatorial
uses of k-mers (via e.g. containment) may be able to resolve strains,
but will need to do so at higher resolution than sourmash's current
parameters. Here approaches such as Agamemnon may be useful.

### A method to evaluate, compare, and study taxonomic lables

The GTDB and NCBI taxonomies are not entirely consonant and our
studies using entropy suggest that a substantial portion of the NCBI
taxonomy is confused. Comparing Table @tbl:gtdb-entropy and Table
@tbl:ncbi-entropy, we see that 5x as many k-mers belong to genomes
that do not share the same family-level labels.  This difference is
due solely to the taxonomic labels. It is not necessarily surprising
that NCBI is so different since GTDB is directly constructed using
content-based phylogeny, but it does suggest that there are many
places where the NCBI taxonomy should be examined closely.

(drill down; contamination, etc.)

### Shannon entropy and unicity on k-mers are robust ways to study, evaluate, and summarize large databases

Exploration of these results suggest that k-mer size and scaled do not
dramatically affect our conclusions. (Confirm me, please :).

## Conclusion

This is easy mode: this kind of taxonomic classification is "just" a
database lookup.  What messes up taxonomic classification with k-mers
is (1) biology (redundancy) and (2) humans (taxonomy). (1) is
resolvable to a significant extent with combinatorics. (2) can be
tackled with better metrics and systematic improvement. Here we
provide measures that assist with both.

Despite this, biological questions remain that are out of scope of
this paper: correctness and completeness of reference databases matters.
And we really also say nothing about generalizability, where we know
that we have problems.
